From cf9f66c12998024833d9ef93ee4df176c1241754 Mon Sep 17 00:00:00 2001
From: Evgeniy Polyakov <zbr@ioremap.net>
Date: Tue, 28 Jun 2016 18:17:00 +0300
Subject: [PATCH] rapidjson: fixed isnan definition with modern compilers
 (xenial g++ 5.3.1)

---
 include/handystats/rapidjson/writer.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/handystats/rapidjson/writer.h b/include/handystats/rapidjson/writer.h
index 78663d9..f4685cd 100644
--- a/include/handystats/rapidjson/writer.h
+++ b/include/handystats/rapidjson/writer.h
@@ -6,7 +6,7 @@
 #include "internal/strfunc.h"
 #include <cstdio>	// snprintf() or _sprintf_s()
 #include <new>		// placement new
-#include <cmath>	// isnan(double)
+#include <cmath>	// std::isnan(double)
 
 #ifdef _MSC_VER
 #pragma warning(push)
@@ -164,7 +164,7 @@ class Writer {
 
 	//! \todo Optimization with custom double-to-string converter.
 	void WriteDouble(double d) {
-		if (isnan(d)) {
+		if (std::isnan(d)) {
 			d = 0;
 		}
 		char buffer[100];
